External Logging link for container nodes

enoodle opened the pull request:
Adding the common logging link support for container nodes.

![gifrecord_2017-03-01_231329](https://cloud.githubusercontent.com/assets/3123328/23483698/b3a6e48e-fedc-11e6-9a67-6481d46f3c65.gif)

enoodle commented:
This depends on https://github.com/ManageIQ/manageiq/pull/13319 (I pulled over the commit from there, will rebase once its merged)

Fryguy commented on app/models/container_node.rb:
```diff
@@ -79,4 +80,19 @@ def ipaddress
     URI::HTTP.build(:host => ipaddress, :port => 9090)
   end
+
+  supports :common_logging do
```
Not sure I like the phrase "common logging" as that is a Red Hat idiom. cc @blomquisg

enoodle replied:
I'll answer on https://github.com/ManageIQ/manageiq/pull/13319

Fryguy commented on config/settings.yml:
```diff
@@ -80,6 +80,8 @@
   :use_vim_broker: true
   :use_vim_broker_ems: true
+:container_logging:
+  :common_logging_route: logging-kibana-ops
```
This feels wrong as a setting in the general settings...feels like it should a per-provider settings or perhaps a special endpoint on each provider.

enoodle replied:
Ill answer on https://github.com/ManageIQ/manageiq/pull/13319

enoodle commented:
I removed the commit from the parent PR so this will only be green after it is merged.

simon3z commented:
@enoodle can you handle the rebase?

simon3z commented:
@enoodle remind me, I think there was no real conclusion on what the kibana search should be for nodes. Is it reliable enough in this PR? cc @portante

portante commented:
@enoodle, this looks like it tries to catch both hostnames and ManageIQ names.  Ideally we'd have all logs from a given cluster tagged with a certain value in the "`tags`" field so that we can query like so to using the shortened hostname, `tags:clusterA AND (hostname:${hostname} OR hostname:$(basename ${hostname})` (syntax here using `bash` for example purposes).

enoodle commented:
@portante I changed to try the hostname from the kubernetes label first. I also changed to make the query itself easier to read.
About the tag: currently it is not available to me, so my only suggestion is that we can add it later when it is more commonly available.

portante commented on app/models/container_node.rb:
```diff
@@ -85,4 +86,20 @@ def evaluate_alert(_alert_id, _event)
+  def external_logging_path
+    hostname = (ipaddress || name || "").split('.').first
+    query = "bool:(filter:(or:!((term:(hostname:#{hostname})),(term:(hostname:#{name})))))"
+    "/#/discover?_g=()&_a=(columns:!(_source),filters:!((meta:(disabled:!f,index:'.operations.*',key:hostname,negate:!f),#{query})),index:'.operations.*',interval:auto,query:(query_string:(analyze_wildcard:!t,query:'*')),sort:!(time,desc))"
```
Can you factor out the `.operations.*` index name as a variable?  It appears in two places in the string.  Also, I tried to use this URL with a few adjustments for names in my Kibana instance, but I get an error.  Are you sure this is right?

Additionally, could we add some fields by default so that it is not just the `_source` field, which is large blob?  Perhaps adding `hostname`, `level`, `pod_name`, and `message` would be sufficient initially.

enoodle replied:
I am not sure that this URL is right, but it is working for me (I have added a gif, made it for the UI Patch but it fits here too).
To get this URL I opened Kibana and told it to filter by hostname. I then changed the query to catch the both fqdn and hostname and trimmed it a little. This is where the `_source` came from. To my understanding after playing with this a little this changes the columns that are displayed next to each message in the list. I will set this to the columns you suggested.
The field `level` is empty for me on all messages but I assume that future improvments with the data model will change that.

portante commented on app/models/container_node.rb:
```diff
@@ -85,4 +86,20 @@ def evaluate_alert(_alert_id, _event)
+
+  supports :external_logging_support do
+    unless ext_management_system.respond_to?(:external_logging_route_name)
+      unsupported_reason_add(:external_logging_support, _('This provider type does not support external_logging_support'))
```
Is this message from the department of the redundancy department?  ;)

enoodle replied:
I will un-robotize this. This message is not for the end user though.

portante commented:
@enoodle, I'll try to get this URL working for me in my kibana instance and report back.

portante commented:
I made a simple python script to verify the base string works now.  Looks good!

```
#!/usr/bin/env python

def external_logging_path(index, hostname_l):
    hostname_s = (hostname_l or "*").split('.')[0]
    query = "bool:(filter:(or:!((term:(hostname:'%(hostname_s)s')),(term:(hostname:'%(hostname_l)s')))))" % { 'hostname_l': hostname_l, 'hostname_s': hostname_s }
    return  "#/discover?_g=()&_a=(columns:!(hostname,level,kubernetes.pod_name,message),filters:!((meta:(disabled:!f,index:%(index)s,key:hostname,negate:!f),%(query)s)),index:%(index)s,interval:auto,query:(query_string:(analyze_wildcard:!t,query:'*')),sort:!('@timestamp',desc))" % { 'index': index, 'query': query }

print external_logging_path("[logstash-]YYYY.MM.DD", "perf44.example.com")
print external_logging_path("[logstash-]YYYY.MM.DD", "gprfs003.example.com")
print external_logging_path("[logstash-]YYYY.MM.DD", "172.17.40.144")
```

simon3z commented:
@enoodle can you re-test this if it actually works for you with OpenShift and Kibana-ops. Then I think we're ready for merge.

enoodle commented:
@simon3z I verified this with the latest update to the UI PR. updated the GIF.

simon3z commented on app/models/container_node.rb:
```diff
@@ -85,4 +86,22 @@ def evaluate_alert(_alert_id, _event)
+
+  def external_logging_path
+    fqdn = (ipaddress || name || "")
+    hostname = fqdn.split('.').first
```
@enoodle @portante if `fqdn` is an ipaddress... wouldn't `hostname` be the first octet?

Also shouldn't you protect against `hostname` possibly being `nil`?

If neither `ipaddress` or `name` are defined why do we continue with `""`? Wouldn't it generate a meaningless query?

enoodle replied:
@simon3z The name of the method was misleading so I changed it from `ipaddress` to `kubernetes_hostname`. If it is indeed returning an IP address instead of an hostname then we will add a search for the first octect (OR the full IP). It will not be optimal but won't crush anything and considering that none of our options gave a valid hostname it seems like an acceptable option.

I fixed `hostname` to have a default value of ""

portante commented on app/models/container_node.rb:
```diff
@@ -72,17 +73,35 @@ def perf_rollup_parents(interval_name = nil)
+  def external_logging_path
+    fqdn = (kubernetes_hostname || name || "")
+    hostname = fqdn.split('.').first || ""
+    query = "bool:(filter:(or:!((term:(hostname:'#{hostname}')),(term:(hostname:'#{fqdn}')))))"
```
@simon3z, unfortunately, the logging subsystem does not return a consistent set of hostnames for the same entity.  So some parts of the logging subsystem gather the FQDN, other parts might use the shortened name.  I believe this query is trying to account for that.  We have to file bugs with the logging subsystem to fix the collectors to ensure an FQDN.

Having said that, data collection is never going to be perfect, so it is likely we'll need to keep this around.

enoodle replied:
@simon3z I want a query to get all the message with hostname that equals to the fqdn of the node OR the hostname of the node. This is derived from looking at the values of this field in the logs that are currently being collected.
I have tested that if either `fqdn` or `hostname` will be empty string `""` it will not add results to the search query results.

simon3z replied:
> I have tested that if either fqdn or hostname will be empty string "" it will not add results to the search query results.

@enoodle as mentioned above I think this should be explicit in the logic that builds the query rather than implicit in the empty string handling of kibana.

enoodle commented:
@simon3z PTAL

simon3z commented on app/models/container_node.rb:
```diff
@@ -72,17 +73,39 @@ def perf_rollup_parents(interval_name = nil)
+    query = "#{query}," if !hostname.blank? && !fqdn.blank?
+    query = "#{query}(term:(hostname:'#{hostname}'))" if !hostname.blank?
+    query = "#{query})))"
+    index = ".operations.*"
```
@enoodle mmmm... logic is still incomplete because `fqdn.split('.').first` could generate an hostname that is `== fqdn`:

```ruby
fqdn = "node_short_hostname"
hostname = fqdn.split('.').first || ""
=> "node_short_hostname"
```

Also you check multiple times (at least 4) for `!fqdn.blank?` and `!hostname.blank?`

Since this search is term-based couldn't you just do something like:

```ruby
node_hostnames = [kubernetes_hostname || name]  # node name cannot be empty, it's an ID
node_hostnames.push(node_hostnames.first.split('.').first)
node_hostnames_query = node_hostnames.uniq.map{|x| "(term:(hostname:'#{x}'))"}.join(",")
...
```

Which for `test.example.com` would generate a `node_hostnames_query = "(term:(hostname:'test.example.com')),(term:(hostname:'test'))"`

Also it seems that the URL may deserve a constant.

simon3z commented on spec/models/manageiq/providers/kubernetes/container_manager/container_node_spec.rb:
```diff
@@ -32,4 +32,30 @@
+    it "queries only for the name/fqdn when hostname can't be parsed" do
+      node = FactoryGirl.create(:container_node, :name => "hello")
+      query = get_query(node.external_logging_path)
+      expect(query).to eq("bool:(filter:(or:!((term:(hostname:'hello')),(term:(hostname:'hello')))))")
```
@enoodle shouldn't this fail now?

enoodle replied:
@simon3z Yes, I am on it. This is not working as intended yet.

enoodle commented:
@simon3z  PTAL

simon3z commented:
@enoodle please add a test that takes advantage of `kubernetes.io/hostname`

simon3z commented:
@miq-bot assign Fryguy

Fryguy commented on app/models/container_node.rb:
```diff
@@ -72,17 +73,36 @@ def perf_rollup_parents(interval_name = nil)
     true
   end
+
+  supports :external_logging_support do
```
This name is funny because it duplicates the work support in certain caller paths.  I think you just want `supports :external_logging`.

enoodle replied:
Right, do you want me to change this name in this PR? This will involve changing it in "app/models/mixins/support_feature_mixin.rb" and "app/models/manageiq/providers/container_manager.rb" too.

durandom replied:
I opt for changing it 👍  But in it's own PR - and check if the UI depends on it

Fryguy replied:
Oh I didn't realize it wasn't introduced in this PR.  Yeah, then separate PR for sure.

Fryguy commented on app/models/container_node.rb:
```diff
@@ -72,17 +73,36 @@ def perf_rollup_parents(interval_name = nil)
+
+  supports :external_logging_support do
+    unless ext_management_system.respond_to?(:external_logging_route_name)
+      unsupported_reason_add(:external_logging_support, _('This provider type does not support Extrenal Logging'))
```
Typo `Extrenal` -> `External`

Fryguy replied:
Does it make sense to make this a dynamic supports method?  Why not just `support_not :external_logging`.  @durandom Thoughts here?

enoodle replied:
@Fryguy This is for the Kubernetes container provider. It doesn't use routes so it can't support this feature.

Fryguy commented on spec/models/manageiq/providers/kubernetes/container_manager/container_node_spec.rb:
```diff
@@ -32,4 +32,37 @@
+  describe "#external_logging_path" do
+    def get_query(path)
+      index = ".operations.*"
+      prefix_len = "/#/discover?_g=()&_a=(columns:!(hostname,level,kubernetes.pod_name,message),filters:!((meta:(disabled:!f,index:'#{index}',key:hostname,negate:!f),".length
```
It feels wrong to duplicate this string in the test.  Is there a way to keep it in the model?  Perhaps a pair of methods for constructing and desconstructing the query string in the model?
